src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -1735,5 +1759,63 @@ private RendezvousAwaitable<int> InternalReceiveResponseHeadersAsync(WinHttpRequ
 
             return state.LifecycleAwaitable;
         }
+
+        internal void ChangeCleanerTimer(TimeSpan timeout)
+        {
+            Debug.Assert(_certificateCleanupTimer != null);
+            if (_certificateCleanupTimer!.Change(timeout, Timeout.InfiniteTimeSpan))
+            {
+                isTimerRunning = timeout != Timeout.InfiniteTimeSpan;
+            }
+        }
+
+        private void ClearStaleCertificates()
+        {
+            Debug.Assert(_certificateCleanupTimer != null);
+
+            foreach (KeyValuePair<IPAddress, CachedCertificateValue> kvPair in cachedCertificates)
+            {
+                if (IsStale(kvPair.Value.LastUsedTime))
+                {
+                    cachedCertificates.TryRemove(kvPair.Key, out CachedCertificateValue _);
+                }
+            }
+
+            ChangeCleanerTimer(cachedCertificates.IsEmpty ? Timeout.InfiniteTimeSpan : s_cleanCachedCertificateTimeout);
+
+            static bool IsStale(long lastUsedTime)
+            {
+                long now = Stopwatch.GetTimestamp();
+                return (now - lastUsedTime) > s_staleTimeout;
+            }
+        }
+
+        internal void CheckTimer()
```

Comment:
The name here isn't very descriptive. IMO these should be two separate methods, e.g. something along the lines of `EnsureCleanupTimerRunning`, `StopCleanupTimerIfEmpty`.

The logic may be easier to follow if we move the logic into a shared location instead of splitting it between the handler and callback files.